readme.md
# NoFlow

A minimal server middleware composer for the future.
It's mostly used to create http server, handle proxy, etc.
The interesting part is that it also works fine without any ES6 or ES7 syntax,
it's up to you to decide how fancy it will be. And because its middlewares are just normal
functions, they can be easily composed with each other.

To use noflow, you only have to remember a single rule "Any async function should and will return a Promise".

# Features

- Super lightweight, only one dependency, 200 sloc, learn it in 5 minutes
- Static-typed with typescript
- Faster than Express.js and Koa, see the benchmark section
- Based on Promise, works well with async/await
- Supports almost all exist Express-like middlewares


**For examples, goto folder [examples](examples)**.

To run the examples, you have to install the dependencies of this project: `npm i`.
Such as, to run the [examples/basic.js](examples/basic.js), use command like:
`node_modules/.bin/noe examples/basic.js`

[![NPM version](https://badge.fury.io/js/noflow.svg)](http://badge.fury.io/js/noflow) [![Build Status](https://travis-ci.org/ysmood/noflow.svg)](https://travis-ci.org/ysmood/noflow) [![Build status](https://ci.appveyor.com/api/projects/status/github/ysmood/noflow?svg=true)](https://ci.appveyor.com/project/ysmood/noflow) [![Deps Up to Date](https://david-dm.org/ysmood/noflow.svg?style=flat)](https://david-dm.org/ysmood/noflow)

# Quick Start

Install it: `npm i noflow`.

### Hello World Example

```javascript
import flow from "noflow";

let app = flow();

// Everything pushed into the app will be converted to a
// middleware object sanely, even it's a string, buffer or anything else.
app.push("hello world");

app.listen(8123);
```

### ES5

Without ES7, you can still have all the good stuffs of Noflow.

```javascript
var flow = require("noflow").default;

var app = flow();

app.push(function ($) {
    return $.next().then(function () {
        console.log("done");
    });
});

app.push(function ($) {
    $.body = "hello world";
});

app.listen(8123);
```

### ES7

Designed for the future ES7.

```javascript
import flow from "noflow";

let app = flow();

app.push(

    async ({ next }) => {
        await next();
        console.log("done");
    },

    $ => $.body = "hello world"

);

app.listen(8123);
```

### Routes

You can write routes quickly by using select interface of [NoKit](https://github.com/ysmood/nokit).

```javascript
import flow from "noflow";
import kit from "nokit";
let { match, select } = kit.require("proxy");

let app = flow();

app.push(
    select("/test", $ => $.body = $.url),

    select(
        // Express.js like url selector.
        match("/items/:id"),
        $ => $.body = $.url.id
    ),

    select(
        {
            url: "/api",
            method: /GET|POST/ // route both GET and POST
        },
        $ => $.body = $.method + " " + $.url
    ),

    select(
        {
            // route some special headers
            headers: {
                host: "a.com"
            }
        },
        $ => $.body = "ok"
    )
);

app.listen(8123);
```

### Express middleware

It's easy to convert an express middleware to noflow middleware.

```javascript
import flow from "noflow";
import bodyParser from "body-parser";

let app = flow();

let convert = (h) => ({ req, res, next }) =>
    new Promise((resolve, reject) =>
        h(req, res, (err) => {
            if (err)
                return reject(err);
            else
                return next().then(resolve);
        })
    );

app.push(
    convert(bodyParser.json()),
    $ => {
        $.body = $.req.body;
    }
);

app.listen(8123);
```

# API

If you use typescript, you don't have to read the API below.

- ## **[flow](src/index.ts?source#L62)**

    Create an array instance with some handy server helper methods.

    - **<u>example</u>**:

        ```js
        import flow from "noflow"
        let app = flow();
        app.push("OK");
        app.listen(8123).then(() => console.log("started"));
        ```




# Status code

Noflow will auto handle status code `200`, `204`, `404` and `500` for you only if
you haven't set the status code yourself.

- `204`: If you don't set `$.body`, such as it's `null` or `undefined`.

- `404`: If no middleware found.

- `500`: If any middleware reject or throws an error.

- `userDefined`: If user set the `$.res.statusCode` manually.

- `200`: If none of the above happens.


# Error handling

A middleware can catch all the errors of the middlewares after it.

With ES5, you can use it like normal promise error handling:

```js
var flow = require("noflow").default;

var app = flow();

app.push(function ($) {
    return $.next().catch(function (e) {
        $.body = e;
    });
});

app.push(function () {
    throw "error";
});

app.push(function () {
    // Same with the `throw "error"`
    return Promise.reject("error");
});

app.listen(8123);
```

With ES7, you can use try-catch directly:

```js
import flow from "noflow";

let app = flow();

app.push(async ($) => {
    try {
        await $.next();
    } catch (e) {
        $.body = e;
    }
});

app.push(() => {
    throw "error";
});

app.listen(8123);
```


# [NoKit](https://github.com/ysmood/nokit)

Noflow relies on the async nature of Promise, when you need async io tools, nokit will be the best choice.
nokit has all the commonly used IO functions with Promise support.

For example you can use them seamlessly:

```js
import flow from "noflow";
import kit from "nokit";
let { select } = kit.require("proxy");

let app = flow();

app.push(
    select(
        "/a",
        kit.readJson("a.json") // readJson returns a Promise
    ),

    select("/b", async $ => {
        let txt = await kit.readFile("b.txt");
        let data = await kit.request("http://test.com/" + $.url);
        $.body = txt + data;
    })
);

app.listen(8123).then(() => {
    kit.request('127.0.0.1:8123/a').then(kit.logs);
});
```

# Benchmark

These comparisons only reflect some limited truth, no one is better than all others on all aspects.
You can run it yourself in terminal: `npm run no -- benchmark`.

```
Node v5.4.0
The less the better:
noflow: 1839.333ms
koa: 2598.171ms
express: 3239.013ms
```


doc/changelog.md
# Changelog

- v0.7

  - **API CHANGE** For CommonJS require, you have to use `require("noflow").default`

- v0.6.0

  - **API CHANGE** the `listener` api is removed
  - fix: the `server` bug

- v0.5.6

  - upd: deps

- v0.5.1

  - init


src/flow.ts
/// <reference path="../typings/node.d.ts" />

"use strict";

/*
    For the sake of performance don't use `let` key word here.
 */

import { isFunction, isArray } from "./utils";
import * as http from "http";
import { Stream } from "stream";
import Promise, { Thenable } from "yaku";

let { STATUS_CODES } = http;


export interface Context {
    /**
     * It will be auto set as the response body.
     */
    body: String | Buffer | Stream | Thenable<any> | Object;

    /**
     * An IncomingMessage object is created by http.Server or http.
     * ClientRequest and passed as the first argument to the 'request' and 'response' event respectively.
     * It may be used to access response status, headers and data.
     * https://nodejs.org/api/http.html#http_http_incomingmessage
     */
    req: http.IncomingMessage;

    /**
     * This object is created internally by a HTTP server--not by the user. It is passed as the second parameter to the 'request' event.
     * The response implements the Writable Stream interface.
     * https://nodejs.org/api/http.html#http_class_http_serverresponse
     */
    res: http.ServerResponse;

    /**
     * It returns a promise which settles after all the next middlewares are setttled.
     */
    next: () => Promise<any>;

}

export interface MiddlewareFn {
    ($: Context): Thenable<any> | any;
}

export type Middleware = MiddlewareFn | Object;

export interface FlowHandler extends MiddlewareFn {
    (req: http.IncomingMessage, res: http.ServerResponse): Promise<any>;
}


/**
 * A promise based function composer.
 * @example
 * Noflow encourages composition.
 * ```js
 * import flow from "noflow"
 * let app = flow();
 * let c = 0;
 * app.push(
 *     $ => $.next(c++),
 *     flow(
 *         $ => $.next(c++),
 *         flow(
 *             $ => $.next(c++),
 *             $ => $.next(c++)
 *         )
 *     ),
 *     $ => $.body = c
 * );
 * app.listen(8123);
 * ```
 */
let flow = function(middlewares: Array<Middleware>): FlowHandler {
    return function(req, res?) {
        let $: Context, parentNext, next;

        // If it comes from a http listener, else it comes from a sub noflow.
        if (res) {
            $ = { req: req, res: res, body: null, next: null };
        } else {
            $ = req;
            parentNext = $.next;

            req = $.req;
            res = $.res;
        }

        let index = 0;

        // Wrap the next middleware.
        next = $.next = function() {
            let mid = middlewares[index++];
            if (mid === undefined) {
                // TODO: #4
                if (parentNext) {
                    return parentNext();
                } else {
                    return Promise.resolve(error404($));
                }
            }

            let ret = tryMid(ensureMid(mid), $);

            // Check if the fn has thrown error.
            if (ret === tryMid) {
                return Promise.reject(tryMidErr);
            } else {
                return Promise.resolve(ret);
            }
        };

        // Begin the initial middleware.
        let promise = next();

        // The root middleware will finnally end the entire $ peacefully.
        if (!parentNext) {
            return promise
                .then(function() { return endCtx($); })
                .then(undefined, function(err) { return errorAndEndCtx(err, $); });
        }

        return promise;
    };
};

// Convert anything to a middleware function.
function ensureMid(mid: Middleware) {
    if (isFunction(mid)) return <MiddlewareFn>mid;

    return function($: Context) { $.body = mid; };
}

// for better performance, hack v8.
let tryMidErr;
function tryMid(fn: MiddlewareFn, $: Context): Thenable<any> | typeof tryMid {
    try {
        return fn($);
    } catch (err) {
        tryMidErr = err;
        return tryMid;
    }
}

function endRes($: Context, data, isStr?: boolean) {
    let buf;
    if (isStr) {
        buf = new Buffer(data);
    } else {
        buf = data;
    }

    if (!$.res.headersSent) {
        $.res.setHeader("Content-Length", buf.length);
    }

    $.res.end(buf);
}

function setStatusCode(res: http.ServerResponse, code: number) {
    if (res.statusCode === 200) res.statusCode = code;
}

function endEmpty(res: http.ServerResponse) {
    setStatusCode(res, 204);
    res.end();
}

function endCtx($: Context) {
    let body = $.body;
    let res = $.res;

    switch (typeof body) {
        case "string":
            endRes($, body, true);
            break;

        case "object":
            if (body == null) {
                endEmpty(res);
            } else if (body instanceof Stream) {
                body.pipe(res);
            } else if (body instanceof Buffer) {
                endRes($, body);
            } else if (isFunction((<Thenable<any>>body).then)) {
                return (<Thenable<any>>body).then(function(data) {
                    $.body = data;
                    return endCtx($);
                });
            } else {
                if (!$.res.headersSent) {
                    res.setHeader("Content-Type", "application/json");
                }
                endRes($, JSON.stringify(body), true);
            }
            break;

        case "undefined":
            endEmpty(res);
            break;

        default:
            endRes($, body.toString(), true);
            break;
    }
}

function errorAndEndCtx(err: Error | any, $: Context) {
    setStatusCode($.res, 500);

    if (process.env.NODE_ENV === "production") {
        $.body = STATUS_CODES[$.res.statusCode];
    } else {
        // print the error details
        if (err instanceof Error)
            $.body = err.stack;
        else
            $.body = err + "";
    }

    // end the context
    return endCtx($);
}

function error404($: Context) {
    setStatusCode($.res, 404);
    $.body = STATUS_CODES[$.res.statusCode];
}

export default function(middlewares: Array<Middleware>) {
    // Make sure we pass in an array
    if (!isArray(middlewares)) {
        middlewares = [].slice.call(arguments);
    }

    return flow(middlewares);
};
